File: migrate/src/java/com/tacitknowledge/util/migration/jdbc/AutoPatchService.java
/* Copyright 2007 Tacit Knowledge LLC
 * 
 * Licensed under the Tacit Knowledge Open License, Version 1.0 (the "License");
 * you may not use this file except in compliance with the License. You may
 * obtain a copy of the License at http://www.tacitknowledge.com/licenses-1.0.
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tacitknowledge.util.migration.jdbc;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import javax.sql.DataSource;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

import com.tacitknowledge.util.migration.MigrationException;

/**
 * Creates an AutoPatch environment using a configuration supplied by dependency
 * injection. Exports a hook that can be called to execute AutoPatch after configuration
 *
 * @author Scott Askew (scott@tacitknowledge.com)
 */
public class AutoPatchService extends JdbcMigrationLauncherFactory
{
    /** Class logger */
    private static Log log = LogFactory.getLog(AutoPatchService.class);
    
    /** The name of the schema to patch */
    private String systemName = null;
    
    /** The data source used to patch the schema */
    private DataSource dataSource = null;
    
    /** The type of database */
    private String databaseType = null;
    
    /** The path to the SQL patches */
    private String patchPath = null;
    
    /** The patch to the post-patch tasks */
    private String postPatchPath = null;
    
    /** Whether we really want to apply the patches, or just look */
    private boolean readOnly = false;
    
    /** A set of contexts, in case you want multi-node patches */
    private List contexts = new ArrayList();

    /**
     * Patches the database, if necessary.
     * 
     * @throws MigrationException if an unexpected error occurs
     */
    public void patch() throws MigrationException
    {
        JdbcMigrationLauncher launcher = getLauncher();
        
        try
        {
            log.info("Applying patches....");
            int patchesApplied = launcher.doMigrations();
            log.info("Applied " + patchesApplied + " "
                + (patchesApplied == 1 ? "patch" : "patches") + ".");
        }
        catch (MigrationException e)
        {
            throw new MigrationException("Error applying patches", e);
        }
    }

    /**
     * Configure and return a JdbcMigrationLauncher to use for patching
     * 
     * @return JdbcMigrationLauncher configured from injected properties
     */
    public JdbcMigrationLauncher getLauncher()
    {
        JdbcMigrationLauncher launcher = getJdbcMigrationLauncher();
        
        // If no one has added a collection of contexts to the service,
        // then take the single-context property configuration and set it in
        if (contexts.size() == 0)
        {
            launcher.addContext(getContext());
        }
        // otherwise, add the collection of contexts in there
        else 
        {
            for (Iterator i = contexts.iterator(); i.hasNext(); )
            {
                launcher.addContext((JdbcMigrationContext)i.next());
            }
        }
        launcher.setPatchPath(getPatchPath());
        launcher.setPostPatchPath(getPostPatchPath());
        launcher.setReadOnly(isReadOnly());
        return launcher;
    }

    /**
     * Configure and return a DataSourceMigrationContext from this object's
     * injected properties
     * 
     * @return DataSourceMigrationContext configured from injected properties
     */
    protected DataSourceMigrationContext getContext()
    {
        DataSourceMigrationContext context = getDataSourceMigrationContext();
        context.setSystemName(getSystemName());
        context.setDatabaseType(new DatabaseType(getDatabaseType()));
        context.setDataSource(getDataSource());
        return context;
    }

    /**
     * @return Returns the dataSource.
     */
    public DataSource getDataSource()
    {
        return dataSource;
    }
    
    /**
     * @param dataSource The dataSource to set.
     */
    public void setDataSource(DataSource dataSource)
    {
        this.dataSource = dataSource;
    }
    
    /**
     * @return Returns the systemName.
     */
    public String getSystemName()
    {
        return systemName;
    }
    
    /**
     * @param systemName The systemName to set.
     */
    public void setSystemName(String systemName)
    {
        this.systemName = systemName;
    }
    
    /**
     * @return Returns the databaseType.
     */
    public String getDatabaseType()
    {
        return databaseType;
    }
    
    /**
     * @param dialect The databaseType to set.
     */
    public void setDatabaseType(String dialect)
    {
        this.databaseType = dialect;
    }
    
    /**
     * @return Returns the patchPath.
     */
    public String getPatchPath()
    {
        return patchPath;
    }
    
    /**
     * @param patchPath The patchPath to set.
     */
    public void setPatchPath(String patchPath)
    {
        this.patchPath = patchPath;
    }
    
    /**
     * @return Returns the postPatchPath.
     */
    public String getPostPatchPath()
    {
        return postPatchPath;
    }
    
    /**
     * @param postPatchPath The postPatchPath to set.
     */
    public void setPostPatchPath(String postPatchPath)
    {
        this.postPatchPath = postPatchPath;
    }

    /**
     * See if we are actually applying patches, or if it is just readonly
     * 
     * @return boolean true if we will skip application
     */
    public boolean isReadOnly()
    {
        return readOnly;
    }

    /**
     * Set whether or not to actually apply patches
     * 
     * @param readOnly boolean true if we should skip application
     */
    public void setReadOnly(boolean readOnly)
    {
        this.readOnly = readOnly;
    }

    /**
     * Get the list of database contexts for multi-node configuration
     * 
     * @return List of JdbcMigrationContext objects for multi-node, or empty List
     */
    public List getContexts()
    {
        return contexts;
    }

    /**
     * Set the list of database contexts for multi-node configuration
     * 
     * @param contexts List of JdbcMigrationContext objects for multi-node, or empty list
     */
    public void setContexts(List contexts)
    {
        this.contexts = contexts;
    }
    
    /**
     * Add a database context to the list of database contexts for multi-node configuration
     *
     * @param context the JdbcMigrationContext to use for multi-node patch application
     */
    public void addContext(JdbcMigrationContext context)
    {
        contexts.add(context);
    }
}

File: migrate/dotnet/AutopatchNet/src/com/tacitknowledge/util/migration/ado/AutoPatchService.cs
/* 
 * Copyright 2007 Tacit Knowledge LLC
 * 
 * Licensed under the Tacit Knowledge Open License, Version 1.0 (the "License");
 * you may not use this file except in compliance with the License. You may
 * obtain a copy of the License at http://www.tacitknowledge.com/licenses-1.0.
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#region Imports
using System;
using log4net;
using log4net.Config;
using MigrationException = com.tacitknowledge.util.migration.MigrationException;
#endregion

namespace com.tacitknowledge.util.migration.ado
{
	
	/// <summary> Creates an AutoPatch environment using a configuration supplied by dependency
	/// injection. Exports a hook that can be called to execute AutoPatch after configuration
	/// 
	/// </summary>
	/// <author>  Scott Askew (scott@tacitknowledge.com)
	/// </author>
	public class AutoPatchService:ADOMigrationLauncherFactory
    {

        #region Member Variables
        /// <summary>Class logger </summary>
        private static ILog log;

        /// <summary>The name of the schema to patch </summary>
        private System.String systemName = null;

        /// <summary>The type of database </summary>
        private System.String databaseType = null;

        /// <summary>The path to the SQL patches </summary>
        private System.String patchPath = null;

        /// <summary>The patch to the post-patch tasks </summary>
        private System.String postPatchPath = null;
        #endregion

        /// <summary> Configure and return a ADOMigrationLauncher to use for patching
		/// 
		/// </summary>
		/// <returns> ADOMigrationLauncher configured from injected properties
		/// </returns>
		/// <exception cref="MigrationException">if there is a problem setting the context
		/// </exception>
		virtual public ADOMigrationLauncher Launcher
		{
			get
			{
				ADOMigrationLauncher launcher = ADOMigrationLauncher;
				launcher.Context = Context;
				launcher.PatchPath = PatchPath;
				launcher.PostPatchPath = PostPatchPath;
				return launcher;
			}
			
		}
		/// <summary> Configure and return a DataSourceMigrationContext from this object's
		/// injected properties
		/// 
		/// </summary>
		/// <returns> DataSourceMigrationContext configured from injected properties
		/// </returns>
		 protected internal DataSourceMigrationContext Context
		{
			get
			{
				DataSourceMigrationContext context = DataSourceMigrationContext;
				context.setSystemName(SystemName);
				context.setDatabaseType(new DatabaseType(DatabaseType));
				
				return context;
			}
			
		}
		
		//UPGRADE_NOTE: Respective javadoc comments were merged.  It should be changed in order to comply with .NET documentation conventions. "ms-help://MS.VSCC.v80/dv_commoner/local/redirect.htm?index='!DefaultContextWindowIndex'&keyword='jlca1199'"
		/// <returns> Returns the systemName.
		/// </returns>
		/// <param name="systemName">The systemName to set.
		/// </param>
		virtual public System.String SystemName
		{
			get
			{
				return systemName;
			}
			
			set
			{
				this.systemName = value;
			}
			
		}
		//UPGRADE_NOTE: Respective javadoc comments were merged.  It should be changed in order to comply with .NET documentation conventions. "ms-help://MS.VSCC.v80/dv_commoner/local/redirect.htm?index='!DefaultContextWindowIndex'&keyword='jlca1199'"
		/// <returns> Returns the databaseType.
		/// </returns>
		/// <param name="dialect">The databaseType to set.
		/// </param>
		virtual public System.String DatabaseType
		{
			get
			{
				return databaseType;
			}
			
			set
			{
				this.databaseType = value;
			}
			
		}
		//UPGRADE_NOTE: Respective javadoc comments were merged.  It should be changed in order to comply with .NET documentation conventions. "ms-help://MS.VSCC.v80/dv_commoner/local/redirect.htm?index='!DefaultContextWindowIndex'&keyword='jlca1199'"
		/// <returns> Returns the patchPath.
		/// </returns>
		/// <param name="patchPath">The patchPath to set.
		/// </param>
		virtual public System.String PatchPath
		{
			get
			{
				return patchPath;
			}
			
			set
			{
				this.patchPath = value;
			}
			
		}
		//UPGRADE_NOTE: Respective javadoc comments were merged.  It should be changed in order to comply with .NET documentation conventions. "ms-help://MS.VSCC.v80/dv_commoner/local/redirect.htm?index='!DefaultContextWindowIndex'&keyword='jlca1199'"
		/// <returns> Returns the postPatchPath.
		/// </returns>
		/// <param name="postPatchPath">The postPatchPath to set.
		/// </param>
		virtual public System.String PostPatchPath
		{
			get
			{
				return postPatchPath;
			}
			
			set
			{
				this.postPatchPath = value;
			}
			
		}
		
		/// <summary> Patches the database, if necessary.
		/// 
		/// </summary>
		/// <throws>  MigrationException if an unexpected error occurs </throws>
		public virtual void  patch()
		{
			ADOMigrationLauncher launcher = Launcher;
			
			try
			{
				log.Info("Applying patches....");
				int patchesApplied = launcher.doMigrations();
				log.Info("Applied " + patchesApplied + " " + (patchesApplied == 1?"patch":"patches") + ".");
			}
			catch (MigrationException e)
			{
				throw new MigrationException("Error applying patches", e);
			}
		}
		static AutoPatchService()
		{
			log = LogManager.GetLogger(typeof(AutoPatchService));
		}
	}
}